[MPSInductor] Add `masked` implementation

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #143966
* #144170
* __->__ #144084
* #144083
* #144162
* #144167

More or less borrowed from 
https://github.com/pytorch/pytorch/blob/22580f160e9ff6f5a54bc5abd03ba3eb75519e10/torch/_inductor/codegen/halide.py#L549-L563

`pytest test/inductor/test_torchinductor.py -k _mps` score is 408 failed, 347 passed, 32 skipped

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

malfet committed:
[MPSInductor] Add trivial `masked` implementation

Which is the same as where for now

[ghstack-poisoned]

malfet committed:
Update on "[MPSInductor] Add trivial `masked` implementation"


Which is the same as where for now

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

malfet committed:
Update on "[MPSInductor] Add trivial `masked` implementation"


Which is the same as where for now

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

malfet committed:
Update on "[MPSInductor] Add trivial `masked` implementation"


Which is the same as where for now

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/mps.py:
```diff
@@ -79,6 +79,11 @@ def index_expr(expr: sympy.Expr, dtype: torch.dtype) -> str:
 
+    @staticmethod
+    def masked(mask: CSEVariable, body: sympy.Expr, other: CSEVariable) -> str:
+        # TODO: Add a proper implementation considering there are no lambdas in Metal
```
We don't actually codegen lambdas here.  For triton/halide this gets codegened to adding masking to any loads inside the body() function.  On CPU it becomes an `if`.

malfet committed:
Update on "[MPSInductor] Add trivial `masked` implementation"


Which is the same as where for now

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

malfet committed:
Update on "[MPSInductor] Add trivial `masked` implementation"


Which is the same as where for now

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

malfet committed:
Update on "[MPSInductor] Add trivial `masked` implementation"


Which is the same as where for now

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

malfet committed:
Update on "[MPSInductor] Add trivial `masked` implementation"


Which is the same as where for now

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

malfet committed:
Update on "[MPSInductor] Add trivial `masked` implementation"


Which is the same as where for now

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

malfet committed:
Update on "[MPSInductor] Add `masked` implementation"


More or less borrowed from 
https://github.com/pytorch/pytorch/blob/22580f160e9ff6f5a54bc5abd03ba3eb75519e10/torch/_inductor/codegen/halide.py#L549-L563

`pytest test/inductor/test_torchinductor.py -k _mps` score is 408 failed, 347 passed, 32 skipped

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

malfet committed:
Update on "[MPSInductor] Add `masked` implementation"


More or less borrowed from 
https://github.com/pytorch/pytorch/blob/22580f160e9ff6f5a54bc5abd03ba3eb75519e10/torch/_inductor/codegen/halide.py#L549-L563

`pytest test/inductor/test_torchinductor.py -k _mps` score is 408 failed, 347 passed, 32 skipped

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -f "Lint + MPS tests are green"
